timemachine/fe/single_topology.py

```diff
@@ -382,141 +359,180 @@ def get_num_connected_components(num_atoms: int, bonds: Collection[Tuple[int, in
-    # either 0 or 180, since the normalized chiral volume is still smooth wrt perturbations
-    all_proper_dummy_chiral_atom_idxs = []
-    all_proper_dummy_chiral_atom_params = []
-    for (c, i, j, k), p in zip(all_dummy_chiral_atom_idxs, all_dummy_chiral_atom_params):
-        ci = canonicalize_bond((c, i))
-        cj = canonicalize_bond((c, j))
-        ck = canonicalize_bond((c, k))
-        if ci in canon_mol_c_bond_idxs_set and cj in canon_mol_c_bond_idxs_set and ck in canon_mol_c_bond_idxs_set:
-            all_proper_dummy_chiral_atom_idxs.append((c, i, j, k))
-            all_proper_dummy_chiral_atom_params.append(p)
-        else:
-            warnings.warn(f"Chiral Volume {c, i, j, k} has a disabled bond, turning off.", ChiralVolumeDisabledWarning)
-
-    mol_c_chiral_atom_idxs = list(mol_a_chiral_atom_idxs) + list(all_proper_dummy_chiral_atom_idxs)
-    mol_c_chiral_atom_params = np.concatenate([mol_a_chiral_atom.params, all_proper_dummy_chiral_atom_params])
-
-    # canonicalize bonds
-    mol_c_bond_idxs_canon = np.array([canonicalize_bond(idxs) for idxs in mol_c_bond_idxs])
-    bond_potential = HarmonicBond(mol_c_bond_idxs_canon).bind(np.array(mol_c_bond_params))
-
-    # chiral atoms need special code for canonicalization, since triple product is invariant
-    # under rotational symmetry (but not something like swap symmetry)
-    canon_chiral_atom_idxs = []
-    for idxs in mol_c_chiral_atom_idxs:
-        canon_chiral_atom_idxs.append(canonicalize_chiral_atom_idxs(idxs))
+    mol_b_top = topology.BaseTopology(mol_b, ff)
+    mol_b_bond_params, mol_b_hb = mol_b_top.parameterize_harmonic_bond(ff.hb_handle.params)
+    mol_b_chiral_atom, _ = mol_b_top.setup_chiral_restraints(
+        DEFAULT_CHIRAL_ATOM_RESTRAINT_K, DEFAULT_CHIRAL_BOND_RESTRAINT_K
+    )
 
-    chiral_atom_idxs = np.array(canon_chiral_atom_idxs, dtype=np.int32).reshape((-1, 4))
-    mol_c_chiral_bond_idxs_canon = [canonicalize_bond(idxs) for idxs in mol_a_chiral_bond_idxs]
-    chiral_bond_idxs = np.array(mol_c_chiral_bond_idxs_canon, dtype=np.int32).reshape((-1, 4))
-    chiral_bond_signs = np.array(mol_a_chiral_bond.potential.signs)
+    def setup_end_state_harmonic_bond_and_chiral_potentials(
+        core: NDArray,
+        a_to_c: NDArray,
+        b_to_c: NDArray,
+        dummy_groups: Dict[int, FrozenSet[int]],
+    ):
+        all_dummy_bond_idxs_, all_dummy_bond_params_ = [], []
+        all_dummy_chiral_atom_idxs_, all_dummy_chiral_atom_params_ = [], []
+
+        for anchor, dg in dummy_groups.items():
+            all_idxs, all_params = setup_dummy_bond_and_chiral_interactions(
+                mol_b_hb.idxs,
+                mol_b_bond_params,
+                mol_b_chiral_atom.potential.idxs,
+                np.asarray(mol_b_chiral_atom.params),
+                dg,
+                anchor,
+                core[:, 1],
+                verify,
+            )
+            # append idxs
+            all_dummy_bond_idxs_.extend(all_idxs[0])
```

Comment:
nit: In some places you have added `np.concatenate` does replacing these extends with `np.concatenate` not improve performance?

Reply:
Good call; I wouldn't think concatenating arrays on the fly would be efficient because it involves unnecessary copying, but I think that appending instead of extending in the loop and calling `np.concatenate` at the end should be faster (and does appear to shave a few more percent). https://github.com/proteneer/timemachine/pull/1360/commits/e3596cd43f4a2f2670467482aa61ce5e9c371a00